### PR TITLE
feat(icons): add support for font-icons using ligatures

### PR DESCRIPTION
**Pull request by ThomasBurleson**
Support for `md-font-library=""` attribute
Improved documentation
Extra demo using Material Design icons.

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@robertmesserle, @jelbourn  - ready for your review.

File: src/components/icon/iconDirective.js [hunk 1]
```diff
@@ -17,44 +17,144 @@ angular.module('material.components.icon', [
...
+
+    // If using font-icons, transclude the ligature or NRCs
+    var tmpl = hasAttrValue('mdFontIcon')    ? '<span class="md-font {{classNames}}" ng-class="fontIcon"></span>' :
+               hasAttrValue('mdFontLibrary') ? '<span ng-transclude></span>' : '';
+
+    // Transpose the mdFontLibrary name to the list of classnames
+    var names = attrValue('mdFontLibrary')  + ' ' +  attrValue('class');
+    element.attr('class',names.trim());
+
+    return $interpolate( tmpl )({ classNames: names.trim() });
```
Comment (robertmesserle) [inline src/components/icon/iconDirective.js]: Using $interpolate here seems a bit heavy-handed - why not just include `names.trim()` in the original template string?

File: src/components/icon/iconDirective.js [hunk 2]
```diff
@@ -17,44 +17,144 @@ angular.module('material.components.icon', [
...
     },
     restrict: 'E',
+    transclude:true,
     template: getTemplate,
     link: postLink
   };
 
   function getTemplate(element, attr) {
-    return attr.mdFontIcon ? '<span class="md-font" ng-class="fontIcon"></span>' : '';
+    var hasAttrValue = function(key) { return angular.isDefined(attr[key]) && attr[key].length };
```
Comment (robertmesserle) [inline src/components/icon/iconDirective.js]: Totally nitpicking, but it seems simpler to use `return attr[key] && attr[key].length`

**Comment by robertmesserle** [comment]
@ThomasBurleson LGTM, just made some minor comments.

File: src/components/icon/iconDirective.js (hunk 1 above)
Comment (ThomasBurleson) [inline src/components/icon/iconDirective.js]: At one point, the tokenized params were more complicated. I will adjust L151.

File: src/components/icon/iconDirective.js (hunk 2 above)
Comment (ThomasBurleson) [inline src/components/icon/iconDirective.js]: agreed

File: src/components/icon/demoFontIconsWithLigatures/index.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<div ng-controller="DemoCtrl" ng-cloak >
+
+  <p>
+    Display 4 Material Design font-icons using ligatures [instead of CSS names]; each with different sizes and colors<br/>
```
Comment (jelbourn) [inline src/components/icon/demoFontIconsWithLigatures/index.html]: Why does a `<p>` end with a `<br>`?

File: src/components/icon/demoFontIconsWithLigatures/script.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  .controller('DemoCtrl', function( $scope ) {
+      // Specify a list of font-icons with ligatures and color overrides
+      var iconData = [
+            {name: 'accessibility'  , color: "#777" },
+            {name: 'question_answer', color: "rgb(89, 226, 168)" },
+            {name: 'backup'         , color: "#A00" },
+            {name: 'email'          , color: "#00A" }
+          ];
+
+      $scope.fonts = [].concat(iconData);
```
Comment (jelbourn) [inline src/components/icon/demoFontIconsWithLigatures/script.js]: Why the `.concat`? iconData isn't used anywhere else, so assigning it directly to the `$scope` would be equivalent.

File: src/components/icon/demoFontIconsWithLigatures/script.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+      // Create a set of sizes...
+      $scope.sizes = [
+        {size:"md-18",padding:0},
+        {size:"md-24",padding:2},
+        {size:"md-36",padding:6},
+        {size:"md-48",padding:10}
```
Comment (jelbourn) [inline src/components/icon/demoFontIconsWithLigatures/script.js]: Space between colon and value

File: src/components/icon/demoFontIconsWithLigatures/script.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+
+angular
+  .module('appDemoFontIconsWithLigatures', ['ngMaterial'])
+  .controller('DemoCtrl', function( $scope ) {
+      // Specify a list of font-icons with ligatures and color overrides
+      var iconData = [
+            {name: 'accessibility'  , color: "#777" },
+            {name: 'question_answer', color: "rgb(89, 226, 168)" },
+            {name: 'backup'         , color: "#A00" },
+            {name: 'email'          , color: "#00A" }
```
Comment (jelbourn) [inline src/components/icon/demoFontIconsWithLigatures/script.js]: Can you consistently use either single-quotes or double-quotes? (I personally prefer single).

File: src/components/icon/iconDirective.js [hunk 7]
```diff
@@ -17,44 +17,146 @@ angular.module('material.components.icon', [
  * @restrict E
  *
  * @description
- * The `md-icon` directive is an markup element useful for showing an icon based on a font-face
- * or a SVG. Both external SVGs (via URLs) or cached SVG from icon sets can be
- * easily loaded and used.
+ * The `<md-icon />` directive is an markup element useful for showing an icon based on a font-icon
```
Comment (jelbourn) [inline src/components/icon/iconDirective.js]: Custom tags can't be self-closing.

File: src/components/icon/iconDirective.js [hunk 8]
```diff
@@ -17,44 +17,146 @@ angular.module('material.components.icon', [
...
+ * or a SVG. Icons are view-only elements that should not be used directly as buttons; instead nest a `<md-icon />`
```
Comment (jelbourn) [inline src/components/icon/iconDirective.js]: "an <md-icon>"

File: src/components/icon/iconDirective.js [hunk 9]
```diff
@@ -17,44 +17,146 @@ angular.module('material.components.icon', [
...
+ * inside a `md-button` to add hover and click features.
  *
+ * When using SVGs, both external SVGs (via URLs) or sets of SVGs [from icon sets] can be
+ * easily loaded and used.When use font-icons, developers must following three (3) simple steps:
```
Comment (jelbourn) [inline src/components/icon/iconDirective.js]: Missing space

File: src/components/icon/iconDirective.js [hunk 10]
```diff
@@ -17,44 +17,146 @@ angular.module('material.components.icon', [
...
     link: postLink
   };
 
   function getTemplate(element, attr) {
-    return attr.mdFontIcon ? '<span class="md-font" ng-class="fontIcon"></span>' : '';
+    var hasAttrValue = function(key) { return attr[key] && attr[key].length      };
+    var attrValue    = function(key) { return hasAttrValue(key) ? attr[key] : '' };
+
+    // If using font-icons, transclude the ligature or NRCs
+    var tmpl = hasAttrValue('mdFontIcon')    ? '<span class="md-font {{classNames}}" ng-class="fontIcon"></span>' :
```
Comment (jelbourn) [inline src/components/icon/iconDirective.js]: Are you amenable to keep lines under 100 characters?

File: src/components/icon/iconDirective.js [hunk 11]
```diff
@@ -17,44 +17,146 @@ angular.module('material.components.icon', [
...
+               hasAttrValue('mdFontLibrary') ? '<span ng-transclude></span>' : '';
+
+    // Transpose the mdFontLibrary name to the list of classnames
+    // For example, Material Icons expects classnames like `.material-icons.md-48` instead of `.material-icons .md-48`
+
+    var names = (attrValue('mdFontLibrary')  + ' ' +  attrValue('class')).trim();
+    element.attr('class',names);
+
+    return $interpolate( tmpl )({ classNames: names });
```
Comment (jelbourn) [inline src/components/icon/iconDirective.js]: Some parens have inner spaces and some don't; it should be consistent. No spaces would be following the Google JS style guide.

**Comment by MigFerreira** [comment]
Hi, 

I cant seem to get the icon visible inside a button. 

codepen: http://codepen.io/MigFerreira/pen/RPpapj

**Comment by shyndman** [comment]
Looks like `letter-spacing: 0.1em` on `.md-button` kills the ligatures.

@ThomasBurleson I'm going to ask the Fonts team to make a change to the served CSS. Might take a day or two.

**Comment by shyndman** [comment]
OK, Fonts team just submitted a CL with the fix. I've asked them to expedite a push to prod.

**Comment by shyndman** [comment]
@MigFerreira Try your demo again. Should work properly.

**Comment by MigFerreira** [comment]
@shyndman Thanks for fast action, looking great!

**Comment by ThomasBurleson** [comment]
Thx @shyndman.
@marcysutton  - do you recall why we have this in typography.scss:
```css
.md-button {
  letter-spacing: 0.010em;
}
```

**Comment by marcysutton** [comment]
@ThomasBurleson it was for the "tracking" section in the Typography spec (aka letter-spacing): https://www.google.com/design/spec/style/typography.html#typography-other-typographic-guidelines

**Pull request merged**